src/dmd/compiler.d

```diff
@@ -109,7 +109,7 @@ struct Compiler
      * as another type for use in CTFE.
      * This corresponds roughly to the idiom *(Type *)&e.
      */
-    extern (C++) static Expression paintAsType(Expression e, Type type)
+    extern (C++) static UnionExp paintAsType(Expression e, Type type)
```

Comment:
Wouldn't `void paintAsType(UnionExp* pue, /*...*/)` be a better signature rather than relying on nrvo?

Though I'm not sure that nrvo would happen, as some compiler backends implementation this function in C++.

Also, C++ headers require updating.

Reply:
NRVO has been working since I invented it 25 years ago. It's not some new thing, everybody does it and it's expected.

Why would C++ call a function meant only for CTFE?

Reply:
Also, quite a bit of dmd relies on this pattern already. See constfold.d.

Reply:
I was forced to change it because of the C++ header.

Reply:
> Why would C++ call a function meant only for CTFE?

The other way around, D calls a function implemented in C++.  This is because `real_t` is a synthetic type in gdc, and only gcc routines know about the internal layout to reinterpret it as another type.

Reply:
For IntegerExp, the `dinteger_t` type could also be converted into a synthetic type in the future for supporting 128-bit (or wider) integral types at compile-time.